src/hotspot/os/linux/os_linux.cpp

```diff
@@ -2941,6 +2950,32 @@ void os::pd_free_memory(char *addr, size_t bytes, size_t alignment_hint) {
   }
 }
 
+size_t os::pd_pretouch_memory(void* first, void* last, size_t page_size) {
+  const size_t len = pointer_delta(last, first, sizeof(char)) + page_size;
+  // Use madvise to pretouch on Linux when THP is used, and fallback to the
+  // common method if unsupported. THP can form right after madvise rather than
+  // being assembled later.
+  if (HugePages::thp_mode() == THPMode::always || UseTransparentHugePages) {
+    int err = 0;
+    if (UseMadvPopulateWrite &&
+        ::madvise(first, len, MADV_POPULATE_WRITE) == -1) {
+      err = errno;
+    }
+    if (!UseMadvPopulateWrite || err == EINVAL) { // Not to use or not supported
+      // When using THP we need to always pre-touch using small pages as the
+      // OS will initially always use small pages.
+      return os::vm_page_size();
+    } else if (err != 0) {
+      log_warning(gc, os)("::madvise(" PTR_FORMAT ", " SIZE_FORMAT
+                          ", %d) failed; error='%s' (errno=%d)",
+                          p2i(first), len, MADV_POPULATE_WRITE,
+                          os::strerror(err), err);
```

Comment:
What other things can go wrong here beside missing kernel support?

Unconditional log output (with log_warning) is tricky. Many tools parse the JVM output and are thrown off by unexpected content. That's why we restrict log_warning to the small band of "stuff that can go wrong at a customer but it is so severe we really need to tell the customer right now".

Stuff that should never go wrong should be assert()ed, or possibly guarantee()'d.

Stuff that can go wrong but is not as severe, should be warned about at a lower level. 

In this case, output may get flooded with warnings if we continue running the VM and repeat the pretouch attempts with other areas.

Reply:
There are some cases shown in madvise_populate at linux/mm/madvise.c. But I'm not sure how to trigger them. So I would decrease the log level here.